third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -44,6 +44,50 @@ int getWmmaVersion(StringRef archGen) {
   return 0;
 }
 
+// Check if the result of current tl.dot is used as the operand(0)
+// of another tl.dot
+bool isChainDotHead(tt::DotOp &dotOp) {
+  auto filter = [&dotOp](Operation *op) {
+    return op->getParentRegion() == dotOp->getParentRegion();
+  };
+  ForwardSliceOptions fwdOpt;
+  fwdOpt.filter = filter;
+  SetVector<mlir::Operation *> fwdSlices;
+  getForwardSlice(dotOp.getResult(), &fwdSlices, fwdOpt);
+  for (Operation *op : fwdSlices) {
+    // ensure output of the first dot is the operand 0 of the second dot
+    if (isa<tt::DotOp>(op) && (op != dotOp)) {
+      auto dOp = dyn_cast<tt::DotOp>(op);
```

Comment:
`op` should never be `dotOp` here given we don't set `fwdOpt.inclusive` to `true`? To be certain you can put an `assert` in the above. Also typically in MLIR we do like

```c++
if (auto userDotOp = dyn_cast<tt::DotOpInterface>()) { ... }
```
Then in the middle you don't need to `cast` again.